ursa_sharing/src/pedersen.rs

```diff
@@ -0,0 +1,232 @@
+            shamir.get_shares_and_polynomial(rng, &blinding)?;
+
+        let mut commitments = Vec::with_capacity(self.limit);
+        for i in 0..self.threshold {
+            let mut g_i = R::zero();
+            g_i.add_assign(&g);
+            g_i.mul_assign(&secret_polynomial.coefficients[i]);
+
+            let mut h_i = R::zero();
+            h_i.add_assign(&h);
+            h_i.mul_assign(&blinding_polynomial.coefficients[i]);
+
+            g_i.add_assign(&h_i);
+
+            commitments.push(ShareVerifier {
+                value: g_i,
+                phantom: PhantomData,
+            });
+        }
+
+        Ok(PedersenVssResult {
+            blinding,
+            blinding_shares,
+            secret_shares,
+            verifier: PedersenVerifier { g, h, commitments },
+        })
+    }
+
+    /// Checks if the share is valid according to verifier
+    pub fn verify_share<S: Field<S>, R: Field<S>>(
+        &self,
+        share: &ShamirShare,
+        blind_share: &ShamirShare,
+        verifier: &PedersenVerifier<S, R>,
+    ) -> SharingResult<()> {
+        let s = S::from_bytes(&share.value)?;
+        if !s.is_valid() {
+            return Err(SharingError::ShareInvalidValue);
+        }
+        let t = S::from_bytes(&blind_share.value)?;
+        if !t.is_valid() {
+            return Err(SharingError::PedersenBlindShareInvalid);
+        }
+
+        let x = S::from_usize(share.identifier);
+        let mut i = S::one();
+
+        // FUTURE: execute this sum of products
+        // c_0 * c_1^i * c_2^{i^2} ... c_t^{i^t}
+        // as a constant time operation using <https://cr.yp.to/papers/pippenger.pdf>
+        // or Guide to Elliptic Curve Cryptography book,
+        // "Algorithm 3.48 Simultaneous multiple point multiplication"
+        // without precomputing the addition but still reduces doublings
+
+        // c_0
+        let mut rhs = R::zero();
+        rhs.add_assign(&verifier.commitments[0].value);
+        for v in &verifier.commitments[1..] {
+            // i *= x
+            i.mul_assign(&x);
+            // c_0 * c_1^i * c_2^{i^2} ... c_t^{i^t}
+            let mut c = R::zero();
+            c.add_assign(&v.value);
+            c.mul_assign(&i);
+            rhs.add_assign(&c);
+        }
+
+        let mut g = R::zero();
+        g.add_assign(&verifier.g);
+        g.negate();
+        g.mul_assign(&s);
+        rhs.add_assign(&g);
+
+        let mut h = R::zero();
+        h.add_assign(&verifier.h);
+        h.negate();
+        h.mul_assign(&t);
+        rhs.add_assign(&h);
+
+        if rhs.is_zero() {
+            Ok(())
+        } else {
+            Err(SharingError::ShareInvalidValue)
+        }
+    }
+
+    /// Reconstruct a secret from shares created from `split_secret`. The shares should be verified
+    /// first by calling `verify_share`. This method assumes all the shares have been verified.
+    /// Usually `verify_share` is called when the share is received.
```

Comment:
Is there a reason not to automatically call verify_share when we are computing combine_shares?  I am worried that this is a misuse resistance issue--this puts the onus on the programmer to remember to call verify_share when the shares are received.

Maybe the default should be to automatically verify all of the shares, although we could have an option to turn this off?

Reply:
It's not terribly time-consuming to call verify_share for each one but I did want to benchmark them separately. Perhaps it's worth the overhead.